### PR TITLE
Support for a docker_build type of doc_independent build

### PR DESCRIPTION
**Pull request by hidmic**
Connected to #568. This pull request adds support for a new kind of package independent documentation builds, tailored to work with Jekyll/Middleman -generated static sites and relying on source side Docker containers to perform the actual build.

### CODE REVIEW

File: ros_buildfarm/config/doc_build_file.py [hunk 1]
```diff
@@ -17,7 +17,9 @@
 DOC_TYPE_ROSDOC = 'rosdoc_lite'
 DOC_TYPE_MANIFEST = 'released_manifest'
 DOC_TYPE_MAKE = 'make_target'
-DOC_TYPES = [DOC_TYPE_ROSDOC, DOC_TYPE_MANIFEST, DOC_TYPE_MAKE]
+DOC_TYPE_EXTERNAL = 'external_site'
```
Comment (nuclearsandwich) [inline ros_buildfarm/config/doc_build_file.py]: I'm not sure if "external site" is best name for this doc type. The other types all describe the build mechanism not the built artifact. I would have expected something more like "docker_container" to similarly describe the how not the what.

Since we'll need to document the new documentation job type on https://github.com/ros-infrastructure/ros_buildfarm/blob/a9f0554840dda39272d76ff19f67c1675d77f8d6/doc/jobs/doc_jobs.rst#different-types-of-documentation that will probably be a good way to settle on a type name. It should match the description in documentation.

File: ros_buildfarm/doc_job.py [hunk 2]
```diff
@@ -407,34 +408,49 @@ def configure_doc_independent_job(
...
-        'upload_root': build_file.upload_root,
 
         'notify_emails': build_file.notify_emails,
 
         'timeout_minutes': build_file.jenkins_job_timeout,
-
-        'credential_id': build_file.upload_credential_id,
     }
+
+    if build_file.documentation_type == 'make_target':
```
Comment (nuclearsandwich) [inline ros_buildfarm/doc_job.py]: I feel somewhat hesitant to tie the upload mechanism to the documentation type. It's possible that an external_site would still publish via ssh/rsync or that we may want to publish make_target docs via git.

This will need further discussion because I think ideally we'd preserve support for the existing upload configuration while deprecating it in favor of a nested structure that we can validate.

Something like below for the git publisher
```yaml
upload:
  type: git
  repository_url: git@github.com:ros-infrastructure/doc-target.git
  repository_branch: gh-pages
  credential_id: <credential_id>
```

and for the doc_independent job that uses make_target
```yaml
upload:
  type: rsync_ssh
  host: some-docs-server.ros.org
  root: /var/www/docs
  user: docsbot
  credential_id: <credential_id>
```

File: ros_buildfarm/templates/doc/doc_independent_site_job.xml.em [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+))@
+@[end if]@
+@(SNIPPET(
+    'property_requeue-job',
+))@
+  </properties>
+@(SNIPPET(
+  'scm_git',
+   url=upload_repository_url,
+   branch_name=upload_repository_branch or 'gh-pages',
```
Comment (nuclearsandwich) [inline ros_buildfarm/templates/doc/doc_independent_site_job.xml.em]: Is there a reason to prefer setting a default in the template rather than setting it earlier on or not setting a default and asserting?

File: ros_buildfarm/templates/doc/doc_independent_site_job.xml.em [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+))@
+  <assignedNode>@(node_label)</assignedNode>
+  <canRoam>false</canRoam>
+  <disabled>false</disabled>
+  <blockBuildWhenDownstreamBuilding>false</blockBuildWhenDownstreamBuilding>
+  <blockBuildWhenUpstreamBuilding>false</blockBuildWhenUpstreamBuilding>
+  <triggers>
+@(SNIPPET(
+    'trigger_timer',
+    spec='H 3 * * *',
```
Comment (nuclearsandwich) [inline ros_buildfarm/templates/doc/doc_independent_site_job.xml.em]: How long does the test job take and how often do we want it to run?

File: ros_buildfarm/config/doc_build_file.py (hunk 1 above)
Comment (hidmic) [reply ros_buildfarm/config/doc_build_file.py]: I don't have a strong preference, but the current job is somewhat coupled to the deploy mechanism as a git commit happens (and it is expected to happen) within the container.

File: ros_buildfarm/doc_job.py (hunk 2 above)
Comment (hidmic) [reply ros_buildfarm/doc_job.py]: Makes perfect sense and I do like it better. However, if we go down that route, no repository management can take place inside the container. Not even a git commit.

In any case, I fear that improvement may delay this new functionality reaching production. What do you think?

File: ros_buildfarm/templates/doc/doc_independent_site_job.xml.em (hunk 3 above)
Comment (hidmic) [reply ros_buildfarm/templates/doc/doc_independent_site_job.xml.em]: It is actually being set before. A leftover, I'll remove it.

File: ros_buildfarm/templates/doc/doc_independent_site_job.xml.em (hunk 4 above)
Comment (hidmic) [reply ros_buildfarm/templates/doc/doc_independent_site_job.xml.em]: I presume it'll take somewhere between one and two hours on a good network connection. And I understand we want this to be built on a daily basis (just as it is for other independent doc types). We did not discuss any other form of triggering.

**Comment by hidmic** [comment]
@nuclearsandwich Just pushed minimal changes to trigger the discussion. If we get (really) picky:

1. You're right about fully decoupling build and deploy stages. No need for `make` or a containerized version of it to know anything about the upload.
2. Using 'external_site' is then a poor choice.
3. Having both `make_target` and `external_site` types hooked up of the same job logic is weird. Those would look better decoupled.
4. If all of the above holds, then source side containers could even be an alternative to the base container provided for the other doc types instead of a separate type. 

I'm open for changes, but I'd rather give up flexibility to have this up and running.

File: ros_buildfarm/doc_job.py (hunk 2 above)
Comment (nuclearsandwich) [reply ros_buildfarm/doc_job.py]: As much as I want to take the time to improve things while we're touching them, I think the change is either a pre-requisite to this PR or just an enhancement to follow up on and not something that should be included in this work.

File: ros_buildfarm/templates/doc/doc_independent_site_job.xml.em (hunk 4 above)
Comment (nuclearsandwich) [reply ros_buildfarm/templates/doc/doc_independent_site_job.xml.em]: I had to remind myself of the Jenkins cron spec behavior. This runs once a day at an idle time during 3-4AM. :+1:

**Comment by nuclearsandwich** [comment]
[rendered view of the updated doc](https://github.com/ros-infrastructure/ros_buildfarm/blob/33282d7de17e07ba03b24ee1e4172a296cdacf33/doc/jobs/doc_jobs.rst)

File: ros_buildfarm/templates/doc/doc_independent_site_job.xml.em [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        ' --build-arg user=`whoami` --build-arg uid=`id -u`' +
+        ' -t %s-env .' % doc_repository_name,
+        'echo "# END SECTION"',
+        '',
+        'echo "# BEGIN SECTION: Build & update site - %s"' % doc_repository_name,
+        'docker run' +
+        ' --rm' +
+        ' --net=host' +
+        ' --cidfile=$WORKSPACE/docker_site/docker.cid' +
+        ' -v $WORKSPACE/repository:/tmp/scaffold' +
```
Comment (nuclearsandwich) [inline ros_buildfarm/templates/doc/doc_independent_site_job.xml.em]: Is there a reason the name `scaffold` was chosen here? Was there a reason not to just use repository?

Comment (hidmic) [reply ros_buildfarm/templates/doc/doc_independent_site_job.xml.em]: I was just reusing Jekyll terminology. No strong opinion either, we can use `repository` or any other name you find more appropriate.

File: ros_buildfarm/templates/doc/doc_independent_site_job.xml.em [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        ' -v $WORKSPACE/site:/tmp/site' +
+        ' %s-env update_site /tmp/scaffold /tmp/site' % doc_repository_name,
```
Comment (nuclearsandwich) [inline ros_buildfarm/templates/doc/doc_independent_site_job.xml.em]: I'm thinking about the external_site "API" between the buildfarm scripts and the source-side containers. The API right now is the update_site command with the paths to the scaffold and site directories as positional arguments.

A more "Docker-ish" way to define the API might be to pass the mount points in as environment variables and then let the Dockerfile use that information however it wants by specifying a [CMD](https://docs.docker.com/engine/reference/builder/#cmd) directive in the source Dockerfile.

Doing so would increase the coupling between ros_buildfarm and docker but decrease the coupling between the source-side containers and ros_buildfarm so I think it's a small net gain.

File: ros_buildfarm/templates/doc/doc_independent_site_job.xml.em (hunk 5 above)
Comment (nuclearsandwich) [reply ros_buildfarm/templates/doc/doc_independent_site_job.xml.em]: Thinking more broadly on this, could we use `upload_repository`to describe the repository that holds the published content and `doc_repository` for the documentation source repository everywhere? The more consistent we are with naming the easier it is to grep through the codebase looking for things and I try to propagate names from the config through to their use as much as is feasible.

File: doc/jobs/doc_jobs.rst [hunk 7]
```diff
@@ -31,6 +31,12 @@ Each *doc build files* can have one of three specific documentation types:
   honor an environment variable identifying the destination of the generated
   documentation.
 
+* the **external_site** type also operates on a single independent repository.
+  For each *doc build file* with that type a Jenkins job is created. Such repository
```
Comment (dirk-thomas) [inline doc/jobs/doc_jobs.rst]: Each sentence should start on a new line (see the [developer guideline](https://github.com/ros2/ros2/wiki/Developer-Guide#markdown)).

Same below.

File: ros_buildfarm/config/doc_build_file.py [hunk 8]
```diff
@@ -17,7 +17,9 @@
...
+DOC_TYPES = [DOC_TYPE_ROSDOC, DOC_TYPE_MANIFEST,
```
Comment (dirk-thomas) [inline ros_buildfarm/config/doc_build_file.py]: To avoid aligning please wrap after the open bracket.

File: ros_buildfarm/config/doc_build_file.py [hunk 9]
```diff
@@ -72,9 +74,16 @@ def __init__(self, name, data):
...
 
-        # doc_repositories can only be used with doc type make_target
+        # doc_repositories can only be used with make_target and external_site
+        # doc types
         is_make_target_type = self.documentation_type == DOC_TYPE_MAKE
-        assert not self.doc_repositories or is_make_target_type
+        is_external_site_type = self.documentation_type == DOC_TYPE_EXTERNAL
+        are_doc_repositories_allowed = (is_make_target_type or
+                                        is_external_site_type)
+        assert not self.doc_repositories or are_doc_repositories_allowed
```
Comment (dirk-thomas) [inline ros_buildfarm/config/doc_build_file.py]: With the logic getting more complex I would suggest to decouple this like:

```
if self.documentation_type not in (DOC_TYPE_MAKE, DOC_TYPE_EXTERNAL):
    assert not self.doc_repositories
```

File: ros_buildfarm/config/doc_build_file.py [hunk 10]
```diff
@@ -72,9 +74,16 @@ def __init__(self, name, data):
...
+
+        # Single doc_repository allowed for external_site doc type
+        assert not is_external_site_type or len(self.doc_repositories) == 1
```
Comment (dirk-thomas) [inline ros_buildfarm/config/doc_build_file.py]: Why is this limitation necessary? Wouldn't it be easy & convenient if multiple repositories could be listed as for doc independent one?

File: ros_buildfarm/doc_job.py [hunk 11]
```diff
@@ -399,6 +399,7 @@ def configure_doc_independent_job(
 
     job_config = _get_doc_independent_job_config(
         config, config_url, job_name, build_file)
+
```
Comment (dirk-thomas) [inline ros_buildfarm/doc_job.py]: Avoid unrelated whitespace changes.

File: ros_buildfarm/doc_job.py [hunk 12]
```diff
@@ -407,34 +408,49 @@ def configure_doc_independent_job(
...
+        template_name = 'doc/doc_independent_site_job.xml.em'
+        job_data.update({
+            'doc_repository_url': build_file.doc_repositories[0],
+            'upload_repository_url': build_file.upload_repository_url,
+            'upload_repository_branch': build_file.upload_repository_branch,
+            'upload_credential_id': build_file.upload_credential_id,
+        })
+    else:
+        raise JobValidationError(
+            "Not independent documentation_type: " +
```
Comment (dirk-thomas) [inline ros_buildfarm/doc_job.py]: Use single quotes.

File: ros_buildfarm/config/doc_build_file.py (hunk 10 above)
Comment (nuclearsandwich) [reply ros_buildfarm/config/doc_build_file.py]: It occurred to me to raise this point as well. It would be reasonable to expect that one job could clone N doc_repositories then build and run N containers, even if the N we need is one.

I didn't raise this question as the use of the jenkins git publisher means that all containers be required to transform the same source repository, but that isn't actually different from all the doc jobs being published to the same rsync destination.

So unless there's something else Dirk and I are missing I think lifting this limitation and iterating over all doc_repositories to build their containers is reasonable.

Comment (hidmic) [reply ros_buildfarm/config/doc_build_file.py]: @nuclearsandwich @dirk-thomas AFAIK, and correct me if I'm wrong, but jenkins git publisher will only allow us to handle one destination repo, that in the case of multiple repos would not be the (most common) case. I'd expect each to commit to the same gh-pages branch in different repositories.

(I'm addressing comments, will reply the rest today)

Comment (nuclearsandwich) [reply ros_buildfarm/config/doc_build_file.py]: I think you're right about the limitation, but the existing independent packages jobs all push to the same rsync target. A set of external sites could potentially all push to subdirectories of the same destination repository and that could be useful.

Comment (hidmic) [reply ros_buildfarm/config/doc_build_file.py]: Hmm, I had not considered that possibility. Alright, fair enough, will add that then.

File: ros_buildfarm/doc_job.py (hunk 12 above)
Comment (hidmic) [reply ros_buildfarm/doc_job.py]: Done.

File: ros_buildfarm/doc_job.py (hunk 11 above)
Comment (hidmic) [reply ros_buildfarm/doc_job.py]: My bad, fixed.

File: ros_buildfarm/config/doc_build_file.py (hunk 9 above)
Comment (hidmic) [reply ros_buildfarm/config/doc_build_file.py]: Good point, done.

File: ros_buildfarm/config/doc_build_file.py (hunk 8 above)
Comment (hidmic) [reply ros_buildfarm/config/doc_build_file.py]: Sure. Done.

File: doc/jobs/doc_jobs.rst (hunk 7 above)
Comment (hidmic) [reply doc/jobs/doc_jobs.rst]: Good catch. Fixed.

File: doc/jobs/doc_jobs.rst [hunk 13]
```diff
@@ -31,6 +31,12 @@ Each *doc build files* can have one of three specific documentation types:
...
+  must provide its own Dockerfile below `docker/image/`. Within that container, an
+  **update_site** command taking the repository path and the site repository path
+  as arguments must be available.
```
[the remark this answers is not in the document]
Comment (hidmic) [reply doc/jobs/doc_jobs.rst]: Documentation updated to match the new environment variable based logic.

File: ros_buildfarm/templates/doc/doc_independent_site_job.xml.em (hunk 6 above)
Comment (hidmic) [reply ros_buildfarm/templates/doc/doc_independent_site_job.xml.em]: I was going to argue against but I just read the CMD directives are overridden by explicitly given arguments. Logic changed.

File: ros_buildfarm/templates/doc/doc_independent_site_job.xml.em (hunk 5 above)
Comment (hidmic) [reply ros_buildfarm/templates/doc/doc_independent_site_job.xml.em]: That sounds perfectly reasonable. Names changed.

**Comment by hidmic** [comment]
@nuclearsandwich @dirk-thomas let me know if there's anything else you'd like to change.

File: ros_buildfarm/doc_job.py [hunk 14]
```diff
@@ -407,34 +407,48 @@ def configure_doc_independent_job(
...
+            build_file.documentation_type
+        )
+
+    repository_args, script_generating_key_files = \
+        get_repositories_and_script_generating_key_files(config=config)
+
+    job_data.update({
+        'script_generating_key_files': script_generating_key_files,
+        'repository_args': repository_args,
+    })
```
[the remark this answers is not in the document]
Comment (hidmic) [reply ros_buildfarm/doc_job.py]: I was trying to avoid duplication, but it's ok anyways, I can roll it back.

File: ros_buildfarm/doc_job.py [hunk 15]
```diff
@@ -407,34 +407,48 @@ def configure_doc_independent_job(
...
     job_data = {
         'job_priority': build_file.jenkins_job_priority,
         'node_label': get_node_label(build_file.jenkins_job_label),
 
-        'ros_buildfarm_repository': get_repository(),
-
         'doc_repositories': build_file.doc_repositories,
-
-        'script_generating_key_files': script_generating_key_files,
+        'ros_buildfarm_repository': get_repository(),
```
[the remark this answers is not in the document]
Comment (hidmic) [reply ros_buildfarm/doc_job.py]: Same as above, though ordering is orthogonal.

**Comment by hidmic** [comment]
@dirk-thomas I'm OK with any naming you'd like. I chose 'external_site' because there is some coupling between the build and deploy phases (i.e. the container commits) but we can call it 'docker_build' or something like that.

File: scripts/doc/generate_doc_independent_job.py [hunk 16]
```diff
@@ -25,6 +25,7 @@
 from ros_buildfarm.config import get_global_doc_build_files
 from ros_buildfarm.config import get_index
 from ros_buildfarm.config.doc_build_file import DOC_TYPE_MAKE
+from ros_buildfarm.config.doc_build_file import DOC_TYPE_EXTERNAL
```
[the remark this answers is not in the document]
Comment (hidmic) [reply scripts/doc/generate_doc_independent_job.py]: Good point.

File: ros_buildfarm/doc_job.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (hidmic) [reply ros_buildfarm/doc_job.py]: `job_data` is populated differently based on the documentation type. I've rearranged things to look more like they used to though. PTAL.

File: doc/jobs/doc_jobs.rst [hunk 17]
```diff
@@ -31,6 +31,14 @@ Each *doc build files* can have one of three specific documentation types:
   honor an environment variable identifying the destination of the generated
   documentation.
 
+* the **docker_build** type also operates on independent repositories.
+  For each *doc build file* with that type a Jenkins job is created.
+  Each repository must provide its own Dockerfile below `docker/image/`,
```
Comment (nuclearsandwich) [inline doc/jobs/doc_jobs.rst]: It'd be great to add an example to the docs. I opted not to permalink it to keep the example up-to-date in theory.

```suggestion
  Each repository must provide its own Dockerfile below `docker/image/` (Example: [ros2/rosindex](https://github.com/ros2/rosindex/tree/ros2/docker/image)),
```

**Comment by nuclearsandwich** [comment]
@dirk-thomas since this is an independent job type do you think we should deploy it from this branch and test before merging?

**Comment by dirk-thomas** [comment]
> I posted one suggestion to add an example to the docs.

I would skip this since in the future this example might change which would likely become outdated in the docs without noticing. The current buildfarm config should be a "good enough" example once this feature is rolled out.

> since this is an independent job type do you think we should deploy it from this branch and test before merging?

Absolutely, good idea.

**Comment by hidmic** [comment]
> The current buildfarm config should be a "good enough" example

True, though there's no example on how the Dockerfile should look like. Documentation is explicit about it, but IMHO an example is always helpful.

**Comment by dirk-thomas** [comment]
Just for the record: http://build.ros.org/job/doc_rosindex/

**Comment by hidmic** [comment]
Updated to better handle the `targets` stanza in doc-build files, as [per request](https://github.com/ros-infrastructure/ros_buildfarm_config/pull/123#discussion_r228202996). @dirk-thomas @nuclearsandwich PTAL.

**Comment by dirk-thomas** [comment]
The documentation for the doc build file needs to be updated to match the logic.

**Comment by hidmic** [comment]
Ready for another round.

File: doc/jobs/doc_jobs.rst [hunk 18]
```diff
@@ -31,6 +31,15 @@ Each *doc build files* can have one of three specific documentation types:
   honor an environment variable identifying the destination of the generated
   documentation.
 
+* the **docker_build** type also operates on independent repositories.
+  For each *doc build file* with that type a Jenkins job is created.
+  Each repository must provide its own Dockerfile below `docker/image/`
+  (Example: [ros2/rosindex](https://github.com/ros2/rosindex/tree/ros2/docker/image)),
```
Comment (chapulina) [inline doc/jobs/doc_jobs.rst]:
```suggestion
  (Example: [ros-infrastructure/rosindex](https://github.com/ros-infrastructure/rosindex/tree/ros2/docker/image)),
```

**Comment by hidmic** [comment]
Thanks for the reviews @nuclearsandwich and @dirk-thomas !

**Comment by dirk-thomas** [comment]
The `external-site-job-support` branch in this repo should probably be deleted and the deployed Jenkins job reconfigured to use the master branch.

**Comment by hidmic** [comment]
Branch is on my fork, but I can drop it if you want. I don't have permission to redeploy the job though, @nuclearsandwich may I ask you to do so?

**Comment by nuclearsandwich** [comment]
> Branch is on my fork, but I can drop it if you want. I don't have permission to redeploy the job though, @nuclearsandwich may I ask you to do so?

I should have clarified that this request was fulfilled.

In order to easily branch-deploy this I pushed a copy of your PR branch to the this fork (ros-infrastructure/ros_buildfarm) and deployed from there. I've since updated the config to use the master branch and deleted that copy of your PR branch. You're welcome to maintain any branches on your fork indefinitely or remove them at your leisure.

File: ros_buildfarm/templates/snippet/publisher_publish-over-git.xml.em [hunk 19]
```diff
@@ -0,0 +1,12 @@
...
+  <pushMerge>false</pushMerge>
+  <pushOnlyIfSuccess>true</pushOnlyIfSuccess>
+  <forcePush>false</forcePush>
+  <branchesToPush>
+    <hudson.plugins.git.GitPublisher_-BranchToPush>
+      <targetRepoName>origin</targetRepoName>
+      <branchName>@branch</branchName>
+    </hudson.plugins.git.GitPublisher_-BranchToPush>
+  </branchesToPush>
+</hudson.plugins.git.GitPublisher>
```
Comment (dirk-thomas) [inline ros_buildfarm/templates/snippet/publisher_publish-over-git.xml.em]: The indentation is wrong resulting in a diff when regenerating the job. See #619 for the fix.
